python/demo/demo_poisson_matrix_free.py

```diff
@@ -0,0 +1,264 @@
+M_fem = fem.form(M, dtype=dtype)
+
+# ### Matrix-free conjugate gradient solver
+#
+# The right hand side vector $b - A x_{\rm bc}$ is the assembly of the linear
+# form $L$ where the essential Dirichlet boundary conditions are implemented
+# using lifting. Since we want to avoid assembling the matrix `A`, we compute
+# the necessary matrix-vector product using the linear form `M` explicitly.
+
+# Apply lifting: b <- b - A * x_bc
+b = fem.assemble_vector(L_fem)
+ui.x.array[:] = 0.0
+fem.set_bc(ui.x.array, [bc], scale=-1.0)
+fem.assemble_vector(b.array, M_fem)
+b.scatter_reverse(la.InsertMode.add)
+
+# Set BC dofs to zero on right hand side
+fem.set_bc(b.array, [bc], scale=0.0)
+b.scatter_forward()
+
+# To implement the matrix-free CG solver using *DOLFINx* vectors, we define the
+# function `action_A` to compute the matrix-vector product $y = A x$.
+
+
+def action_A(x, y):
+    # Set coefficient vector of the linear form M and ensure it is updated
+    # across processes
+    ui.x.array[:] = x.array
+    ui.x.scatter_forward()
+
+    # Compute action of A on ui using the linear form M
+    y.array[:] = 0.0
+    fem.assemble_vector(y.array, M_fem)
+    y.scatter_reverse(la.InsertMode.add)
+
+    # Set BC dofs to zero
+    fem.set_bc(y.array, [bc], scale=0.0)
+
+
+# ### Basic conjugate gradient solver
+#
+# Solves the problem `A x = b`, using the function `action_A` as the operator,
+# `x` as an initial guess of the solution, and `b` as the right hand side
+# vector. `comm` is the MPI Communicator, `max_iter` is the maximum number of
+# iterations, `rtol` is the relative tolerance.
+
+
+def cg(comm, action_A, x: la.Vector, b: la.Vector, max_iter: int = 200, rtol: float = 1e-6):
+    rtol2 = rtol**2
+
+    nr = b.index_map.size_local
+
+    def _global_dot(comm, v0, v1):
+        # Only use the owned dofs in vector (up to nr)
+        return comm.allreduce(np.vdot(v0[:nr], v1[:nr]), MPI.SUM)
+
+    # Get initial y = A.x
+    y = la.vector(b.index_map, 1, dtype)
+    action_A(x, y)
+
+    # Copy residual to p
+    r = b.array - y.array
+    p = la.vector(b.index_map, 1, dtype)
+    p.array[:] = r
+
+    # Iterations of CG
+    rnorm0 = _global_dot(comm, r, r)
+    rnorm = rnorm0
+    for k in range(max_iter):
+        action_A(p, y)
+        alpha = rnorm / _global_dot(comm, p.array, y.array)
+
+        x.array[:] += alpha * p.array
+        r -= alpha * y.array
+        rnorm_new = _global_dot(comm, r, r)
+        beta = rnorm_new / rnorm
+        rnorm = rnorm_new
+        if comm.rank == 0:
+            print(k, np.sqrt(rnorm / rnorm0))
+        if rnorm / rnorm0 < rtol2:
+            x.scatter_forward()
+            return k
+        p.array[:] = beta * p.array + r
+
+    raise RuntimeError(f"Solver exceeded max iterations ({max_iter}).")
+
+
+# This matrix-free solver is now used to compute the finite element solution.
+# The finite element solution's approximation error as compared with the
+# exact solution is measured in the $L_2$-norm.
```

Comment:
The solution selected here is actually exactly represented in the FE space chosen. What we're measuring is the iterative solver error as propagated to the FE solution in the L2 norm. If we chose `u_D` to be something not in $P^3$ then we'd have a combination of the two. Perhaps this should be distinguished?

Reply:
Good point, I think we should use something not in $P^3$.